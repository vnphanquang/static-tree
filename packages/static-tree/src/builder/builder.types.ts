/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { TNode, TNodeData, TNodeInit } from '../node';

import { ExtendedTNodeBuilder } from './builder';

/**
 * @public
 *
 * {@link TNode} extended with children inline as properties, intended to provide
 * idiomatic dot notation parent-to-child node access, so that you can do:
 *
 * ```typescript
 * root.child.grandChild.$.path(); // -\> `root/child/grandChild`
 * ```
 *
 * instead of:
 *
 * ```typescript
 * root.$.children().find(c => c.key === 'child')?.$.children().find(c => c.key === 'grandChild')?.$.path();
 * ```
 *
 * @remarks
 *
 * {@link ExtendedTNode} is constructed behind the scene using {@link tBuild} (preferred)
 * and {@link ExtendedTNodeBuilder}.
 */
export type ExtendedTNode<
  Key extends string = never,
  AncestorKeys extends string = never,
  ChildrenRecord extends Record<string, ExtendedTNode> = {},
  Data extends TNodeData = {},
> = TNode<Data, Key, AncestorKeys> & ChildrenRecord;

/**
 * @public
 *
 * Instruction for building nested {@link ExtendedTNode}.
 *
 */
export interface ExtendedTNodeBuildConfig<
  Key extends string,
  AncestorKeys extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData = {},
> {
  pathResolver?: TNodeInit<Data>['pathResolver'];
  build?: ExtendedTNodeBuildCallback<Key, AncestorKeys, ChildrenRecord, Data>;
}

/**
 * @public
 *
 * Build callback for adding data & nested children
 */
export type ExtendedTNodeBuildCallback<
  Key extends string,
  AncestorKeys extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData = {},
> = (builder: ExtendedTNodeBuilder<Key, AncestorKeys>) => ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, Data>;

/**
 * @public
 *
 * Output of {@link tBuild}
 */
export interface TBuildOutput<
  Key extends string,
  AncestorKeys extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
> {
  /**
   * built node with type safety for children and data
   */
  node: ExtendedTNode<Key, AncestorKeys, ChildrenRecord, Data>;
  /**
   * builder for advanced use cases. See {@link tBuild} for examples
   */
  builder: ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, Data>;
}
