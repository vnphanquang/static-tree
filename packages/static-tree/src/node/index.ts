export * from './node';
export * from './node.types';
