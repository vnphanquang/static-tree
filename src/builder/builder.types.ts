/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { TNode, TNodeData } from '../node';

export type ExtendedTNode<
  ChildrenRecord extends Record<string, ExtendedTNode> = {},
  Data extends TNodeData = {},
> = TNode<Data> & ChildrenRecord;
