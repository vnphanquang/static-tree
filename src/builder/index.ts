export * from './builder';
export * from './builder.types';
