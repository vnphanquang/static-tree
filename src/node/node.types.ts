/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { TNode } from './node';

export interface TNodeInit<D extends Record<string, any> = {}> {
  parent?: TNode;
  data?: D;
}

export interface TNodeGetPathParams {
  separator?: string;
  depth?: number;
}

type DeepRecord<T> = {
  [key: string]: T | DeepRecord<T>;
};

export type Serializable = DeepRecord<
  string | number | boolean | boolean | Array<Serializable> | ReadonlyArray<Serializable>
>;

export interface TNodeSerializeOptions<V extends boolean> {
  dataSerializer?: false | (<D extends Record<string, any> = {}>(node: TNode<D>) => Serializable);
  verbose?: V;
}

export interface MinimalSerializedTNode {
  key: string;
  children: MinimalSerializedTNode[];
  data: Serializable;
}

export interface VerboseSerializedTNode extends MinimalSerializedTNode {
  children: VerboseSerializedTNode[];
  pathSegments: string[];
  depth: number;
  isRoot: boolean;
  path: string;
}

export type SerializedTNode = VerboseSerializedTNode | MinimalSerializedTNode;

export type TNodeData = Record<string, any>;
