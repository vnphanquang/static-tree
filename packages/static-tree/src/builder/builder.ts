/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { TNode } from '../node';
import type { TNodeData, TNodeInit } from '../node';

import type { ExtendedTNode, ExtendedTNodeBuildConfig, TBuildOutput } from './builder.types';

/**
 * @public
 *
 * Builder for {@link ExtendedTNode}
 *
 * @remarks
 *
 * This is used internally by {@link tBuild}. Alternatively, this can
 * be used if a more method-based, verbose solution is preferred.
 *
 * @example
 *
 * ```typescript
 * import { ExtendedTNodeBuilder } from 'static-tree';
 *
 * const node = new ExtendedTNodeBuilder('key')
 *   .addData({ number: 1, boolean: true, string: 'string' })
 *   .addChild('childOne', {
 *      build: (builder) => builder
 *        .addData({ childData: 'something else' })
 *        .addChild('grandChild'),
 *   })
 *   .addChild('childTwo', { build: (builder) => builder.addChild('grandChild') })
 *   .build();
 * ```
 *
 * @example
 *
 * For advanced usage, `ExtendedTNodeBuilder` can also be used
 * as the input in `addChild`, should it provide any benefit.
 *
 * ```typescript
 * import { tBuild, ExtendedTNodeBuilder } from 'static-tree';
 *
 * const externalBuilder = new ExtendedTNodeBuilder('external')
 *  .addData({ some: 'some' })
 *  .addChild('someChild');
 *
 * const { node } = tBuild('root', {
 *   build: (builder) => builder.addChild(externalBuilder),
 * });
 *
 * node.external.someChild.$.path(); // -\> `root/external/someChild`
 * ```
 */
export class ExtendedTNodeBuilder<
  Key extends string,
  AncestorKeys extends string = never,
  ChildrenRecord extends Record<string, ExtendedTNode> = {},
  Data extends TNodeData = {},
> {
  private _node: ExtendedTNode;
  private _key: Key;

  /**
   * @param key - key for the {@link ExtendedTNode} to be built
   * @param parent - parent {@link TNode} to attach to
   */
  constructor(key: Key, options: Omit<TNodeInit<Data>, 'data'> = {}) {
    this._node = new TNode(key, options);
    this._key = key;
  }

  /**
   * add a type-safe child to the {@link ExtendedTNode} to be built
   *
   * @example
   *
   * `addChild` with just a key (no nested children, no internal data)
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild('root', {
   *   build: (builder) => builder.addChild('child'),
   * });
   * ```
   *
   * @example
   *
   * `addChild` with a build callback for nested children & adding data
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild('root', {
   *   build: (builder) => builder
   *     .addChild('child', {
   *        build: (builder) => builder
   *          .addData({ some: 'child data' })
   *          .addChild('grandChildren'),
   *     }),
   * });
   * ```
   *
   * @example
   *
   * `addChild` with external `ExtendedTNodeBuilder` (advanced use case)
   *
   * ```typescript
   * import { tBuild, ExtendedTNodeBuilder } from 'static-tree';
   *
   * const externalBuilder = new ExtendedTNodeBuilder('external');
   *
   * const { node } = tBuild('root', {
   *   build: (builder) => builder.addChild(externalBuilder),
   * });
   * ```
   *
   * @param input - an {@link ExtendedTNodeBuildConfig} instruction
   * @returns this {@link ExtendedTNodeBuilder}
   */
  public addChild<
    ChildKey extends string,
    GrandChildrenRecord extends Record<string, ExtendedTNode>,
    ChildData extends TNodeData = {},
  >(
    builder: ExtendedTNodeBuilder<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>,
  ): ExtendedTNodeBuilder<
    Key,
    AncestorKeys,
    ChildrenRecord & Record<ChildKey, ExtendedTNode<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>>,
    Data
  >;
  public addChild<
    ChildKey extends string,
    GrandChildrenRecord extends Record<string, ExtendedTNode>,
    ChildData extends TNodeData = {},
  >(
    key: ChildKey,
    config?: ExtendedTNodeBuildConfig<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>,
  ): ExtendedTNodeBuilder<
    Key,
    AncestorKeys,
    ChildrenRecord & Record<ChildKey, ExtendedTNode<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>>,
    Data
  >;
  public addChild<
    ChildKey extends string,
    GrandChildrenRecord extends Record<string, ExtendedTNode>,
    ChildData extends TNodeData = {},
  >(
    input: ChildKey | ExtendedTNodeBuilder<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>,
    config?: ExtendedTNodeBuildConfig<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>,
  ): ExtendedTNodeBuilder<
    Key,
    AncestorKeys,
    ChildrenRecord & Record<ChildKey, ExtendedTNode<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>>,
    Data
  > {
    let rKey: ChildKey;
    let node: TNode;

    if (input instanceof ExtendedTNodeBuilder) {
      rKey = input._key;
      node = input.build();
      node.__.setParent(this._node);
    } else {
      rKey = input;
      const options = { parent: this._node, pathResolver: config?.pathResolver };
      if (config?.build) {
        node = config.build(new ExtendedTNodeBuilder(rKey, options)).build();
      } else {
        node = new TNode(rKey, options);
      }
    }

    (this._node as any)[rKey] = node;

    return this as unknown as ExtendedTNodeBuilder<
      Key,
      AncestorKeys,
      ChildrenRecord & Record<ChildKey, ExtendedTNode<ChildKey, AncestorKeys | Key, GrandChildrenRecord, ChildData>>,
      Data
    >;
  }

  /**
   * add type-safe data to the {@link ExtendedTNode} to be built
   *
   * @example
   *
   * Using `tBuild`
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild('root', {
   *   build: (builder) => builder.addData({ some: 'data' }),
   * });
   * ```
   *
   * @example
   *
   * Using `ExtendedTNodeBuilder`
   *
   * @param data - the {@link TNodeData}
   * @returns this {@link ExtendedTNodeBuilder}
   */
  public addData<D extends TNodeData>(data: D): ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, D> {
    this._node.__.setData(data);
    return this as unknown as ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, D>;
  }

  /**
   * Build the {@link ExtendedTNode}
   *
   * @returns node {@link ExtendedTNode}
   */
  public build(): ExtendedTNode<Key, AncestorKeys, ChildrenRecord, Data> {
    return this._node as ExtendedTNode<Key, AncestorKeys, ChildrenRecord, Data>;
  }
}

/**
 * @public
 *
 * Build helper that uses {@link ExtendedTNodeBuilder} internally.
 *
 * @remarks
 *
 * Along with a built {@link ExtendedTNode}, `tBuild` outputs
 * an {@link ExtendedTNodeBuilder} for advanced use cases, see examples
 * for more details.
 *
 * @example
 *
 * Typical usage
 *
 * ```typescript
 * import { tBuild } = 'static-tree';
 *
 * const { node } = tBuild('root', {
 *  build: (builder) => builder
 *    .addData({ some: 'root data' })
 *    .addChild('leaf')
 *    .addChild('child', {
 *      build: (builder) => builder.addChild('grandChild', {
 *        build: (builder) => builder.addData({ some: 'grand child data' }),
 *      }),
 *    }),
 * });
 *
 * node.$.data().some // -\> 'root data';
 * node.child.grandChild.$.data().some // -\> 'grand child data';
 * node.child.grandChild.$.path() // -\> 'root/child/grandChild';
 * ```
 *
 * @example
 *
 * Use `tBuild` builder output in nested `ExtendedTNodeBuildCallback`
 *
 * ```typescript
 * const { builder: externalBuilder } = tBuild('external', {
 *   build: (builder) => builder.addChild('child').addData({ some: 'data' }),
 * });
 *
 * const { node } = tBuild('root', {
 *   build: (builder) => builder.addChild(externalBuilder),
 * });
 * ```
 *
 * @param build - build instruction
 * @param key - key for the {@link ExtendedTNode} to be built
 * @returns
 */
export function tBuild<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
  AncestorKeys extends string = never,
>(
  builder: ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, Data>,
): TBuildOutput<Key, AncestorKeys, ChildrenRecord, Data>;
export function tBuild<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
  AncestorKeys extends string = never,
>(
  key: Key,
  config?: ExtendedTNodeBuildConfig<Key, AncestorKeys, ChildrenRecord, Data>,
): TBuildOutput<Key, AncestorKeys, ChildrenRecord, Data>;
export function tBuild<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
  AncestorKeys extends string = never,
>(
  input: Key | ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, Data>,
  config?: ExtendedTNodeBuildConfig<Key, AncestorKeys, ChildrenRecord, Data>,
): TBuildOutput<Key, AncestorKeys, ChildrenRecord, Data> {
  let builder: ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, Data>;

  if (input instanceof ExtendedTNodeBuilder) {
    builder = input;
  } else {
    const options = {
      pathResolver: config?.pathResolver,
    };
    if (config?.build) {
      builder = config.build(new ExtendedTNodeBuilder(input, options));
    } else {
      builder = new ExtendedTNodeBuilder(input, options);
    }
  }

  return {
    builder,
    node: builder.build(),
  };
}
