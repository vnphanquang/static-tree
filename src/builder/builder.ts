/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { TNode } from '../node';
import type { TNodeData, TNodeInit } from '../node';

import type { ExtendedTNode, ExtendedTNodeBuildInput, TBuildOutput } from './builder.types';

/**
 * @public
 *
 * Builder for {@link ExtendedTNode}
 *
 * @remarks
 *
 * This is used internally by {@link tBuild}. Alternatively, this can
 * be used if a more method-based, verbose solution is preferred.
 *
 * @example
 *
 * ```typescript
 * import { ExtendedTNodeBuilder } from 'static-tree';
 *
 * const node = new ExtendedTNodeBuilder('key')
 *   .addData({ number: 1, boolean: true, string: 'string' })
 *   .addChild({
 *      key: 'childOne',
 *      build: (builder) => builder
 *        .addData({ childData: 'something else' })
 *        .addChildren('grandChild),
 *   })
 *   .addChild('childTwo', builder => builder.addChild('grandChild'))
 *   .build();
 * ```
 *
 * @example
 *
 * For advanced usage, `ExtendedTNodeBuilder` can also be used
 * as the input in `addChild`, should it provide any benefit.
 *
 * ```typescript
 * import { tBuild, ExtendedTNodeBuilder } from 'static-tree';
 *
 * const externalBuilder = new ExtendedTNodeBuilder('external')
 *  .addData({ some: 'some' })
 *  .addChild('someChild');
 *
 * const { node } = tBuild({
 *   key: 'root',
 *   build: (builder) => builder.addChild(externalBuilder),
 * });
 *
 * node.external.someChild.$.path(); // -\> `root/external/someChild`
 * ```
 */
export class ExtendedTNodeBuilder<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode> = {},
  Data extends TNodeData = {},
> {
  private _node: ExtendedTNode;
  private _key: Key;

  /**
   * @param key - key for the {@link ExtendedTNode} to be built
   * @param parent - parent {@link TNode} to attach to
   */
  constructor(key: Key, options: Omit<TNodeInit<Data>, 'data'> = {}) {
    this._node = new TNode(key, options);
    this._key = key;
  }

  /**
   * add a type-safe child to the {@link ExtendedTNode} to be built
   *
   * @example
   *
   * `addChild` with just a key (no nested children, no internal data)
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder.addChild('child'),
   * });
   * ```
   *
   * @example
   *
   * `addChild` with a build callback for nested children & adding data
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder
   *     .addChild({
   *        key: 'child',
   *        build: (builder) => builder
   *          .addData({ some: 'child data' })
   *          .addChild('grandChildren'),
   *     }),
   * });
   * ```
   *
   * @example
   *
   * `addChild` with external `ExtendedTNodeBuilder` (advanced use case)
   *
   * ```typescript
   * import { tBuild, ExtendedTNodeBuilder } from 'static-tree';
   *
   * const externalBuilder = new ExtendedTNodeBuilder('external');
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder.addChildren(externalBuilder),
   * });
   * ```
   *
   * @param input - an {@link ExtendedTNodeBuildInput} instruction
   * @returns this {@link ExtendedTNodeBuilder}
   */
  public addChild<
    ChildKey extends string,
    GrandChildrenRecord extends Record<string, ExtendedTNode>,
    ChildData extends TNodeData = {},
  >(
    input: ExtendedTNodeBuildInput<ChildKey, GrandChildrenRecord, ChildData>,
  ): ExtendedTNodeBuilder<
    Key,
    ChildrenRecord & Record<ChildKey, ExtendedTNode<GrandChildrenRecord, ChildData>>,
    Data
  > {
    let key: ChildKey;
    let node: TNode;

    if (typeof input === 'string') {
      key = input;
      node = new ExtendedTNodeBuilder(key, { parent: this._node }).build();
    } else if (input instanceof ExtendedTNodeBuilder) {
      key = input._key;
      node = input.build();
      node.__.setParent(this._node);
    } else {
      key = input.key;
      node =
        input
          .build?.(
            new ExtendedTNodeBuilder(key, {
              parent: this._node,
              pathResolver: input.pathResolver,
            }),
          )
          .build() ?? new TNode(key, { parent: this._node });
    }

    (this._node as any)[key] = node;

    return this as unknown as ExtendedTNodeBuilder<
      Key,
      ChildrenRecord & Record<ChildKey, ExtendedTNode<GrandChildrenRecord, ChildData>>,
      Data
    >;
  }

  /**
   * add type-safe data to the {@link ExtendedTNode} to be built
   *
   * @example
   *
   * Using `tBuild`
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder.addData({ some: 'data' }),
   * });
   * ```
   *
   * @example
   *
   * Using `ExtendedTNodeBuilder`
   *
   * @param data - the {@link TNodeData}
   * @returns this {@link ExtendedTNodeBuilder}
   */
  public addData<D extends TNodeData>(data: D): ExtendedTNodeBuilder<Key, ChildrenRecord, D> {
    this._node.__.setData(data);
    return this as ExtendedTNodeBuilder<Key, ChildrenRecord, D>;
  }

  /**
   * Build the {@link ExtendedTNode}
   *
   * @returns node {@link ExtendedTNode}
   */
  public build(): ExtendedTNode<ChildrenRecord, Data> {
    return this._node as ExtendedTNode<ChildrenRecord, Data>;
  }
}

/**
 * @public
 *
 * Build helper that uses {@link ExtendedTNodeBuilder} internally.
 *
 * @remarks
 *
 * Along with a built {@link ExtendedTNode}, `tBuild` outputs
 * an {@link ExtendedTNodeBuilder} for advanced use cases, see examples
 * for more details.
 *
 * @example
 *
 * Typical usage
 *
 * ```typescript
 * import { tBuild } = 'static-tree';
 *
 * const { node } = tBuild({
 *  key: 'root',
 *  build: (builder) => builder
 *    .addData({ some: 'root data' })
 *    .addChild('leaf')
 *    .addChild({
 *      key: 'child',
 *      build: (builder) => builder.addChild({
 *        key: 'grandChild',
 *        build: (builder) => builder.addData({ some: 'grand child data' }),
 *      }),
 *    }),
 * });
 *
 * node.$.data().some // -\> 'root data';
 * node.child.grandChild.$.data().some // -\> 'grand child data';
 * node.child.grandChild.$.path() // -\> 'root/child/grandChild';
 * ```
 *
 * @example
 *
 * Use `tBuild` builder output in nested `ExtendedTNodeBuildCallback`
 *
 * ```typescript
 * const { builder: externalBuilder } = tBuild({
 *   key: 'external',
 *   build: (builder) => builder.addChild('child').addData({ some: 'data' }),
 * });
 *
 * const { node } = tBuild({
 *   key: 'root',
 *   build: (builder) => builder.addChild(externalBuilder),
 * });
 * ```
 *
 * @param build - build instruction
 * @param key - key for the {@link ExtendedTNode} to be built
 * @returns
 */
export function tBuild<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
>(
  input: ExtendedTNodeBuildInput<Key, ChildrenRecord, Data>,
): TBuildOutput<Key, ChildrenRecord, Data> {
  let builder: ExtendedTNodeBuilder<Key, ChildrenRecord, Data>;

  if (typeof input === 'string') {
    builder = new ExtendedTNodeBuilder(input);
  } else if (input instanceof ExtendedTNodeBuilder) {
    builder = input;
  } else {
    builder =
      input.build?.(
        new ExtendedTNodeBuilder(input.key, {
          pathResolver: input.pathResolver,
        }),
      ) ?? new ExtendedTNodeBuilder(input.key);
  }

  return {
    builder,
    node: builder.build(),
  };
}
