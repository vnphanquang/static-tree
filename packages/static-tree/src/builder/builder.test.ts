import { test, expect, describe } from 'vitest';

import { TNode } from '../node';

import { ExtendedTNodeBuilder, tBuild } from './builder';

export const TNODE_MOCK_DATA = {
  number: 1,
  string: 'string',
  boolean: true,
  object: {
    nested: 'nested',
  },
  array: [
    {
      id: 'test',
    },
  ],
  specials: {
    date: new Date('2022-08-18'),
    node: new TNode('example'),
    map: new Map(),
    set: new Set(),
    function: () => 2,
  },
} as const;

const { node: sTree } = tBuild('key', {
  pathResolver: () => 'customPathSegment',
  build: (builder) =>
    builder
      .addData(TNODE_MOCK_DATA)
      .addChild('child', {
        build: (builder) =>
          builder
            .addChild('justKey')
            .addChild(new ExtendedTNodeBuilder('external'))
            .addChild('tBuild', {
              build: (builder) =>
                builder
                  .addChild(tBuild('tBuildJustKey').builder)
                  .addChild(tBuild(new ExtendedTNodeBuilder('tBuildWithBuilder')).builder)
                  .addChild(
                    tBuild('tBuildFull', {
                      build: (builder) => builder.addChild('final'),
                    }).builder,
                  ),
            }),
      })
});

describe('tBuild', () => {
  test('data should be typed and returned correctly', () => {
    expect(sTree.$.data()).toMatchObject(TNODE_MOCK_DATA);
    expect(sTree.$.data().number).toBe(TNODE_MOCK_DATA.number);
  });

  test('children should be typed and match', () => {
    expect(sTree.$.key()).toBe('key');
    expect(sTree.child.$.path()).toBe('customPathSegment/child');
    expect(sTree.child.justKey.$.path()).toBe('customPathSegment/child/justKey');
    expect(sTree.child.external.$.path()).toBe('customPathSegment/child/external');
    expect(sTree.child.tBuild.$.path()).toBe('customPathSegment/child/tBuild');
    expect(sTree.child.tBuild.tBuildJustKey.$.path()).toBe(
      'customPathSegment/child/tBuild/tBuildJustKey',
    );
    expect(sTree.child.tBuild.tBuildWithBuilder.$.path()).toBe(
      'customPathSegment/child/tBuild/tBuildWithBuilder',
    );
    expect(sTree.child.tBuild.tBuildFull.$.path()).toBe(
      'customPathSegment/child/tBuild/tBuildFull',
    );
    expect(sTree.child.tBuild.tBuildFull.final.$.path()).toBe(
      'customPathSegment/child/tBuild/tBuildFull/final',
    );
  });

  test('serialization should match', () => {
    expect(sTree.$.serialize({ verbose: true })).toMatchSnapshot();
  });
});

describe('real life example test', () => {
  const { node: api } = tBuild('api', {
    pathResolver: () => 'https://api.domain.example',
    build: (builder) =>
      builder
        .addChild('auth', {
          build: (builder) =>
            builder
              .addChild('logout')
              .addChild('oauth', {
                build: (builder) => builder.addChild('google').addChild('discord'),
                //...
              })
              .addChild('nestedCamelCaseKey', {
                pathResolver: () => 'nested-camel-case-key',
              }),
        })
        .addChild('camelCaseKey', {
          pathResolver: () => 'camel-case-key',
        })
        .addChild(':organization', {
          pathResolver: (_, arg) => arg,
          build: (builder) =>
            builder.addChild(':user', {
              pathResolver: (_, arg) => arg,
            }),
        }),
  });

  test('full path should match', () => {
    expect(api.auth.oauth.google.$.path(), 'https://api.domain.example/auth/oauth/google');
  });
  test('path with positive depth should match', () => {
    expect(api.auth.oauth.google.$.path({ depth: 2 }), 'oauth/google');
  });
  test('path with negative depth should match', () => {
    expect(api.auth.oauth.google.$.path({ depth: -2 }), 'https://api.domain.example/auth');
  });
  test('custom pathResolver should resolve correctly', () => {
    expect(api.camelCaseKey.$.path()).toBe('https://api.domain.example/camel-case-key');
    expect(api.auth.nestedCamelCaseKey.$.path()).toBe(
      'https://api.domain.example/auth/nested-camel-case-key',
    );
  });
  test('dynamic arguments should work', () => {
    expect(api[':organization'][':user'].$.path({
      args: {
        ':organization': 'test-org',
        ':user': 'test-user',
      },
    })).toEqual('https://api.domain.example/test-org/test-user');
  });
});
