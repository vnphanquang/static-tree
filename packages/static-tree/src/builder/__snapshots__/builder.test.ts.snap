// Vitest Snapshot v1

exports[`tBuild > serialization should match 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {},
          "depth": 3,
          "isRoot": false,
          "key": "justKey",
          "path": "customPathSegment/child/justKey",
        },
        {
          "children": [],
          "data": {},
          "depth": 3,
          "isRoot": false,
          "key": "external",
          "path": "customPathSegment/child/external",
        },
        {
          "children": [
            {
              "children": [],
              "data": {},
              "depth": 4,
              "isRoot": false,
              "key": "tBuildJustKey",
              "path": "customPathSegment/child/tBuild/tBuildJustKey",
            },
            {
              "children": [],
              "data": {},
              "depth": 4,
              "isRoot": false,
              "key": "tBuildWithBuilder",
              "path": "customPathSegment/child/tBuild/tBuildWithBuilder",
            },
            {
              "children": [
                {
                  "children": [],
                  "data": {},
                  "depth": 5,
                  "isRoot": false,
                  "key": "final",
                  "path": "customPathSegment/child/tBuild/tBuildFull/final",
                },
              ],
              "data": {},
              "depth": 4,
              "isRoot": false,
              "key": "tBuildFull",
              "path": "customPathSegment/child/tBuild/tBuildFull",
            },
          ],
          "data": {},
          "depth": 3,
          "isRoot": false,
          "key": "tBuild",
          "path": "customPathSegment/child/tBuild",
        },
      ],
      "data": {},
      "depth": 2,
      "isRoot": false,
      "key": "child",
      "path": "customPathSegment/child",
    },
  ],
  "data": {
    "array": [
      {
        "id": "test",
      },
    ],
    "boolean": true,
    "number": 1,
    "object": {
      "nested": "nested",
    },
    "specials": {
      "date": 2022-08-18T00:00:00.000Z,
      "function": [Function],
      "map": Map {},
      "node": TNode {
        "_children": [],
        "_data": {},
        "_depth": 1,
        "_key": "example",
        "_parent": null,
        "_pathResolver": [Function],
      },
      "set": Set {},
    },
    "string": "string",
  },
  "depth": 1,
  "isRoot": true,
  "key": "key",
  "path": "customPathSegment",
}
`;
